DRY up the report filtering logic for the LHN and optionslist

tgolen opened the pull request:
### Details
From the refactoring of the sidebarlinks component, a lot of the filtering logic was left in a non-DRY state. The same logic was repeated in both SidebarUtils and OptionsListUtils. This PR centralizes the logic, and attempts to make it easier to understand and maintain.

### Fixed Issues
$ https://github.com/Expensify/App/issues/11481
$ https://github.com/Expensify/Expensify/issues/228945

### Tests
I took the time to fully cover the previous functionality with unit tests, so there isn't really any testing to do. Just ensure that the unit tests pass!

- [x] Verify that no errors appear in the JS console



### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps
There is so much logic contained in this filtering that it would be very difficult to list out (well over a thousand different combinations). Here are the simple things to check:

1. In the LHN, make sure that reports are ordered by Pinned > Outstanding IOUs > Drafts > the rest of reports
3. Verify that changing any report in those groups will update the LHN properly
2. Verify that new unread chats show up in the LHN

Search for chats
1. Click on the search button
2. Verify that you can search for chats and click on them to access them

- [ ] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/1228807/193347604-59f513ba-4f7d-4db5-be7a-29668181c08c.png">

#### Mobile Web - Chrome
<img width="362" alt="image" src="https://user-images.githubusercontent.com/1228807/193353223-f03cf081-31ea-4a5b-92e2-420da2814e43.png">
<img width="362" alt="image" src="https://user-images.githubusercontent.com/1228807/193353257-c6d9252a-a5ca-45ff-9c65-5ab2935ff57c.png">


#### Mobile Web - Safari
<img width="439" alt="image" src="https://user-images.githubusercontent.com/1228807/193350945-80e6e34e-231a-49d1-b78c-59ee01250d6b.png">
<img width="439" alt="image" src="https://user-images.githubusercontent.com/1228807/193350956-65f7cc69-0f73-4b78-8ec5-93df44f331ac.png">


#### Desktop
<img width="1200" alt="image" src="https://user-images.githubusercontent.com/1228807/193348193-0fcdc0cd-4031-48bc-a93c-5dd9eb8401d3.png">

#### iOS
<img width="439" alt="2022-09-30_14-20-02" src="https://user-images.githubusercontent.com/1228807/193350212-19e7a700-ea61-4317-84e8-6ef3e22c056f.png">
<img width="439" alt="image" src="https://user-images.githubusercontent.com/1228807/193350232-957e8cf0-1354-4feb-b948-39e1044a5670.png">


#### Android
<img width="362" alt="image" src="https://user-images.githubusercontent.com/1228807/193353172-12e49783-49c0-4190-9f17-a976ad04f3b3.png">
<img width="362" alt="image" src="https://user-images.githubusercontent.com/1228807/193353212-0d09e631-2c19-4b7a-ad12-17129a502f73.png">

tgolen committed:
WIP start moving logic into ReportUtils

tgolen committed:
WIP Move more logic into ReportUtils

tgolen committed:
Move the last logic into ReportUtils

tgolen committed:
Revert temporary changes to old logic

tgolen committed:
Move common logic into a utils file

tgolen committed:
Correct import statement of react

tgolen committed:
Rename test to it

tgolen committed:
Move test to the proper suite

tgolen committed:
Add a test for policy expense chat beta

tgolen committed:
WIP with testing policy expense chats

tgolen committed:
Protect against undefined policies

tgolen committed:
Add tests for policy expense chats

tgolen committed:
WIP debug code for tests

tgolen committed:
Protect against betas missing from mocked method

tgolen committed:
Improve code a bit to refer to constants directly

tgolen committed:
Refer to code from test utility

tgolen committed:
Rename utility file

tgolen committed:
Move permissions mock to a file

tgolen committed:
Update react tree query

tgolen committed:
Decrease test to the minimum

tgolen committed:
Decrease test to the minimum

tgolen committed:
Decrease test to the minimum

tgolen committed:
Decrease test to the minimum

tgolen committed:
Decrease test to the minimum

tgolen committed:
Decrease test to the minimum

tgolen committed:
Decrease test to the minimum

tgolen committed:
Decrease test to the minimum

tgolen committed:
Decrease test to the minimum

tgolen committed:
Decrease test to the minimum

tgolen committed:
Clean up comments

tgolen committed:
Reduce test to the minimum

tgolen committed:
Remove unneeded variables

tgolen committed:
Add back in the missing dependency

tgolen committed:
Add a test for user created policy rooms

tgolen committed:
Add a test for policy default rooms

tgolen committed:
Add a test for default policy rooms on free policies

tgolen committed:
Protect against undefined chat room subtitle

tgolen committed:
WIP on testing all boolean combinations

tgolen committed:
Add tests for all boolean combinations

tgolen committed:
Merge branch 'main' into tgolen-clean-reportfiltering

# Conflicts:
#	src/libs/SidebarUtils.js

tgolen committed:
Improve test description

tgolen committed:
Add method docs and export method

tgolen committed:
Implement new filter method

tgolen committed:
Get tests passing with new filter method

tgolen committed:
Remove old filtered report code

tgolen committed:
DRY the logic for filtering reports

tgolen committed:
Remove unused method

tgolen committed:
Remove unused params

tgolen committed:
Remove unused option for prioritizing IOU debts

tgolen committed:
Remove unused option for prioritizeReportsWithDraftComments

tgolen committed:
Remove unused option for sortByAlphaAsc

tgolen committed:
Remove unused option for prioritizePinnedReports

tgolen committed:
Make sure to exclude reports by default

tgolen committed:
Fix ordering by report name

tgolen committed:
Fix recently updated report test

tgolen committed:
WIP on fixing failed tests

tgolen committed:
Finish getting all tests to pass

tgolen committed:
Fix lint errors

tgolen committed:
Fix a broken test with report name

tgolen committed:
Fix a missing constant

tgolen commented on tests/unit/LHNOrderTest.js:
```diff
@@ -152,165 +17,131 @@ const ONYXKEYS = {
-                    [`${ONYXKEYS.COLLECTION.REPORT}1`]: fakeReport1,
-                    [`${ONYXKEYS.COLLECTION.REPORT}2`]: fakeReport2,
-                    [`${ONYXKEYS.COLLECTION.REPORT}3`]: fakeReport3,
-                    [`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}1`]: fakeReport1Actions,
```
I tried to reduce these tests to as little as necessary to still test what we needed to test. The report actions are not used anywhere in the logic, so I just removed them.

tgolen commented:
bump for initial reviews @marcaaron @Gonals

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -22,6 +22,18 @@ Onyx.connect({
+Onyx.connect({
+    key: ONYXKEYS.CURRENTLY_VIEWED_REPORTID,
+    callback: val => currentlyViewedReportID = val,
+});
```
NAB & FYI but I figured out a way to get rid of `currentlyViewedReportID` completely by upgrading some `react-navigation` stuff and am able to pass the reportID from the route and use it as a single source of truth. Not sure if that will affects your changes much but thought you would be interested in that.

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -191,8 +203,13 @@ function getSearchText(report, reportName, personalDetailList, isChatRoomOrPolic
-            Array.prototype.push.apply(searchTerms, chatRoomSubtitle.split(','));
+
+            // When running tests, chatRoomSubtitle can be undefined due to the Localize() stuff being mocked in the tests.
+            // It's OK to ignore this and just add a null check in here to keep code from crashing.
```
I think in the ideal case we would not have comments that reference "the tests" from outside the tests. Like, this comment could be more useful if it mentioned _which_ test needs this so it feels out of context.

Also wondering if we can just mock `translateLocal()` so that it returns an empty string or something? Why is it returning `undefined`?

tgolen replied:
Ah, sorry. I still haven't had time to look into this. I'll create a followup issue. https://github.com/Expensify/App/issues/11585

marcaaron reviewed (commented):
Hey @tgolen, Focusing on some performance issues at the moment, but the changes LGTM. I will let @Gonals take the first crack at this review and jump in after.

marcaaron commented:
Actually, going to unassign myself here. I've got some competing priorities and don't want to be a blocker.

tgolen committed:
Merge branch 'main' into tgolen-clean-reportfiltering

tgolen committed:
Merge branch 'main' into tgolen-clean-reportfiltering

tgolen committed:
Correct a few reportID types

Gonals commented on src/libs/OptionsListUtils.js:
```diff
@@ -191,8 +203,13 @@ function getSearchText(report, reportName, personalDetailList, isChatRoomOrPolic
+            // It's OK to ignore this and just add a null check in here to keep code from crashing.
+            if (chatRoomSubtitle) {
+                Array.prototype.push.apply(searchTerms, chatRoomSubtitle.split(''));
+                Array.prototype.push.apply(searchTerms, chatRoomSubtitle.split(','));
```
We can use a regex here, right? Something like `.split(/[o\s]/)`

Gonals replied:
Thoughts on this, @tgolen?

tgolen replied:
Oh, sorry. I wasn't really following your suggestion. Why is a regex better here?

Gonals replied:
Not a big reason or anything, but just to avoid calling it twice in a row. With regex we can get it done in one go.

tgolen replied:
Ah, got it. I didn't realize it would replace them both. That sounds like a good improvement to me

Gonals commented on src/libs/OptionsListUtils.js:
```diff
@@ -424,42 +441,33 @@ function getOptions(reports, personalDetails, activeReportID, {
-
-    const sortDirection = [sortByAlphaAsc ? 'asc' : 'desc'];
-    let orderedReports = lodashOrderBy(reports, sortProperty, sortDirection);
+    const sortDirection = ['desc'];
```
NAB. It feels a bit weird to me having these constants defined as single-member arrays. Would it be better to make them strings and "arraying" them on the `lodashOrderBy` call?

Gonals commented on src/libs/ReportUtils.js:
```diff
@@ -851,6 +859,108 @@ function isUnread(report) {
+ *
+ * @param {Object} report
+ * @param {String} report.iouReportID
+ * @param {boolean} report.    if (!report || !report.iouReportID || report.hasOutstandingIOU) {
```
? Bad copy-paste?

Gonals reviewed (changes requested):
Looks good! Just a few comments!

Gonals commented on src/libs/ReportUtils.js:
```diff
@@ -851,6 +859,108 @@ function isUnread(report) {
+ * @param {Object} policies
+ * @returns {boolean}
+ */
+function shouldReportBeInOptionList(report, currentlyViewedReportID, isInGSDMode, currentUserLogin, iouReports, betas, policies) {
```
NAB.
I'm not sure I love the betas as a parameter. Could we get them from Onyx?

tgolen committed:
Clean up comment

tgolen committed:
Remove unnecessary variable

tgolen committed:
Remove duplicate param doc

tgolen committed:
Merge branch 'main' into tgolen-clean-reportfiltering

tgolen committed:
Add a test for archived reports being last

tgolen committed:
Add a test for archived reports being last in focus mode

tgolen committed:
Improve the ordering on option list utils

tgolen commented:
@Gonals I've updated the PR with some improved ordering logic so that it only runs the sort function once, and it no longer uses lodash. I've also added unit tests to ensure archived reports are always at the bottom of the LHN no matter which view mode the user is in.

tgolen committed:
Add missing semi-colons

Gonals commented:
> @Gonals I've updated the PR with some improved ordering logic so that it only runs the sort function once, and it no longer uses lodash. I've also added unit tests to ensure archived reports are always at the bottom of the LHN no matter which view mode the user is in.

Looks great! Happy to see `lodashOrderBy` gone 😆 

I just have one small comment that remained unaddressed and i'm all 👍 .
@marcaaron, I know you are pretty busy, but since you remained assigned, do you want to take a final look at this?

tgolen committed:
Combine duplicate split methods into a single method

tgolen committed:
Make them real regex now

marcaaron commented:
>since you remained assigned, do you want to take a final look at this

I tried to un-assign, but I guess once you leave a comment on a GitHub PR you are assigned for LIFE ! 😂

marcaaron commented on src/libs/OptionsListUtils.js:
```diff
@@ -424,114 +439,48 @@ function getOptions(reports, personalDetails, activeReportID, {
+    // - All archived reports should remain at the bottom
+    const orderedReports = _.sortBy(filteredReports, (report) => {
+        if (ReportUtils.isArchivedRoom(report)) {
+            return -Infinity;
```
NAB, but why `-Infinity` here? Would `-1` not work?

marcaaron commented on src/libs/SidebarUtils.js:
```diff
@@ -93,86 +91,44 @@ Onyx.connect({
+    const filteredReportsWithReportName = _.map(filteredReports, (report) => {
+        const personalDetailMap = OptionsListUtils.getPersonalDetailsForLogins(report.participants, personalDetails);
+        return {
+            ...report,
```
NAB, but it might be good to avoid this spread here. If you have tons of reports I would imagine that cloning them all would be less efficient than something like storing the `reportDisplayNames` in a dictionary. Not sure how less efficient, but spread is one of those things you have to watch out for at scale.

tgolen replied:
That's a good call out. Do you want to create a follow-up issue for that? That could probably help for squeaking out some more performance. With this PR, it was mostly just DRYing up the previous code which is why I didn't do too much modification.

tgolen replied:
Oops, I just realized this was not moving existing code, it was new code that I wrote. I'll create the follow-up issue

marcaaron commented on src/libs/SidebarUtils.js:
```diff
@@ -93,86 +91,44 @@ Onyx.connect({
+    // - Regardless of mode, all archived reports should remain at the bottom
+    const orderedReports = _.sortBy(filteredReportsWithReportName, (report) => {
+        if (ReportUtils.isArchivedRoom(report)) {
+            return isInDefaultMode ? -Infinity : 'ZZZZZZZZZZZZZ';
```
NABs, same question about `-Infinity` and why we would use it over `-1`

The `ZZZZZZZZZZZZZ` here also feels like a magic... word (since can't be a magic number 😂)  how did we come up with this and should we add a code comment so that the reasoning is a bit more clear?

marcaaron reviewed (approved):
Changes LGTM. I did have a few questions, but not sure if they are blockers so will leave it up to @tgolen.

Gonals commented:
> @Gonals looks like this was merged without passing tests. Please add a note explaining why this was done and remove the `Emergency` label if this is not an emergency.

No contributor involved!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

marcaaron commented:
@Gonals Just want to draw attention to this recent update on the checklists [here](https://expensify.slack.com/archives/C03TBKRSB/p1664671005145029). I'm not sure whether this has been standardized yet or what actions we are taking to consistently apply "everyone follows the checklist", but we probably should have used it here.

Gonals commented:
> @Gonals Just want to draw attention to this recent update on the checklists [here](https://expensify.slack.com/archives/C03TBKRSB/p1664671005145029). I'm not sure whether this has been standardized yet or what actions we are taking to consistently apply "everyone follows the checklist", but we probably should have used it here.

Oh, thanks for pointing that out! I didn't realize that was the plan going forwards.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3193367486) to staging by @Gonals in version: 1.2.12-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3230773802) to production by @AndrewGable in version: 1.2.12-4 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

situchan commented on src/libs/ReportUtils.js:
```diff
@@ -851,6 +859,107 @@ function isUnread(report) {
+function shouldReportBeInOptionList(report, currentlyViewedReportID, isInGSDMode, currentUserLogin, iouReports, betas, policies) {
+    // Exclude reports that have no data because there wouldn't be anything to show in the option item.
+    // This can happen if data is currently loading from the server or a report is in various stages of being created.
+    if (!report || !report.reportID || !report.participants || _.isEmpty(report.participants)) {
```
There's more case to exclude.
If `reportName` is missing, it isn't still incomplete report and shouldn't show in option item.
More details about the root cause: https://github.com/Expensify/App/issues/29198#issuecomment-1756529588
